crates/qpy/src/circuit_writer.rs

```diff
@@ -0,0 +1,910 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+// Circuit writer module: converts a QuantumCircuit into a qpy file
+
+// We use the following terminology:
+// 1. "Pack": To create a struct (from formats.rs) from the original data
+// 2. "Serialize": To create binary data (Bytes) from the original data
+// 3. "Write": To write to a file obj the serialization of the original data
+// Ideally, serialization is done by packing in a binrw-enhanced struct and using the
+// `write` method into a `Cursor` buffer, but there might be exceptions.
+
+use binrw::Endian;
+use hashbrown::{HashMap, HashSet};
+use indexmap::IndexSet;
+use numpy::ToPyArray;
+
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3::types::{PyAny, PyDict};
+
+use qiskit_circuit::bit::{
+    ClassicalRegister, PyClbit, PyQubit, QuantumRegister, Register, ShareableClbit, ShareableQubit,
+};
+use qiskit_circuit::circuit_data::{CircuitData, CircuitStretchType, CircuitVarType};
+use qiskit_circuit::circuit_instruction::CircuitInstruction;
+use qiskit_circuit::converters::QuantumCircuitData;
+use qiskit_circuit::imports;
+use qiskit_circuit::operations::{ArrayType, Operation, OperationRef};
+use qiskit_circuit::packed_instruction::{PackedInstruction, PackedOperation};
+
+use crate::annotations::AnnotationHandler;
+use crate::bytes::Bytes;
+use crate::formats::{self, InitialLayoutItemV2Pack};
+use crate::params::pack_param_obj;
+use crate::py_methods::{
+    gate_class_name, getattr_or_none, py_get_condition_data_from_inst,
+    py_get_instruction_annotations, py_pack_param, py_pack_pauli_evolution_gate,
+    recognize_custom_operation, serialize_metadata,
+};
+use crate::value::{
+    GenericValue, QPYWriteData, bit_types, circuit_instruction_types, expression_var_declaration,
+    get_circuit_type_key, pack_generic_value, pack_standalone_var, pack_stretch, register_types,
+    serialize,
+};
+
+use crate::UnsupportedFeatureForVersion;
+
+/// packing the qubits and clbits of a specific instruction into CircuitInstructionArgPack
+pub fn get_packed_bit_list(
```

Comment:
```suggestion
// packing the qubits and clbits of a specific instruction into CircuitInstructionArgPack
fn get_packed_bit_list(
```
Most of the functions in this file can be private, so the suggestion here applies to other functions below.